validators/cgroup_validator.go

```diff
@@ -46,15 +46,14 @@ func (c *CgroupsValidator) Validate(spec SysSpec) ([]error, []error) {
 	if err != nil {
 		return nil, []error{errors.Wrap(err, "failed to get cgroup subsystems")}
 	}
-	if err = c.validateCgroupSubsystems(spec.Cgroups, subsystems); err != nil {
-		return nil, []error{err}
-	}
-	return nil, nil
+	errs := c.validateCgroupSubsystems(spec.CgroupSpec.Required, subsystems, true)
+	warns := c.validateCgroupSubsystems(spec.CgroupSpec.Optional, subsystems, false)
+	return warns, errs
 }
 
-func (c *CgroupsValidator) validateCgroupSubsystems(cgroupSpec, subsystems []string) error {
-	missing := []string{}
-	for _, cgroup := range cgroupSpec {
+func (c *CgroupsValidator) validateCgroupSubsystems(cgroups, subsystems []string, required bool) []error {
```

Comment:
Style: might be simpler to merely report back the list of missing cgroups and let Validate above construct and report the appropriate errors and warnings.  Not going to insist on it if you prefer to do it this way, I just think it might mean less code change.

Reply:
Not sure which one i prefer to be honest. Updated to separate the concern now (as a new commit to make it easier to see the diff). Do you think the new one is better?

Reply:
Sorry, didn't see this until just now.  I prefer this, but not strongly enough to want to insist on it.